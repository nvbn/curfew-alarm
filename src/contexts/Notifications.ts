import * as ExpoNotifications from "expo-notifications";
import { createContext } from "react";

import INotifications from "../dependencies/INotifications";

const Notifications = createContext<INotifications>(ExpoNotifications);

export default Notifications;
