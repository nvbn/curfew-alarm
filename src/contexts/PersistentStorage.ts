import { createContext } from "react";

import IPersistentStorage, {
  PersistentStorageDefaultImpl,
} from "../dependencies/IPersistentStorage";

const PersistentStorage = createContext<IPersistentStorage>(
  PersistentStorageDefaultImpl,
);

export default PersistentStorage;
