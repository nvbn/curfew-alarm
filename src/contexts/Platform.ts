import { createContext } from "react";

import IPlatform, { PlatformDefaultImpl } from "../dependencies/IPlatform";

const Platform = createContext<IPlatform>(PlatformDefaultImpl);

export default Platform;
