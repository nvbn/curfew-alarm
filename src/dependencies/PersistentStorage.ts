import AsyncStorage from "@react-native-async-storage/async-storage";
import { createContext } from "react";

/**
 * Interface to access device persistent storage.
 */
export interface IPersistentStorage {
  getItem(key: string): Promise<string | null>;

  setItem(key: string, value: string): Promise<void>;
}

export const PersistentStorageDefaultImpl: IPersistentStorage = AsyncStorage;

export const PersistentStorageCtx = createContext<IPersistentStorage>(
  PersistentStorageDefaultImpl,
);
