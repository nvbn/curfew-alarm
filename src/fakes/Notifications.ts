import INotifications, {
  NotificationContent,
  NotificationPermissions,
  NotificationPushToken,
  NOTIFICATIONS_PERMISSIONS_GRANTED,
} from "../dependencies/INotifications";

type Behavior = {
  scheduleNotificationAsync: Promise<string>;
  getPermissionsAsync: Promise<NotificationPermissions>;
  requestPermissionsAsync: Promise<NotificationPermissions>;
  getExpoPushTokenAsync: Promise<NotificationPushToken>;
  setNotificationChannelAsync: Promise<void>;
};

export const NOTIFICATIONS_DEFAULT_FAKE_TOKEN = "EXPO_PUSH_NOTIFICATIONS_TOKEN";

const defaultBehavior: Behavior = {
  scheduleNotificationAsync: Promise.resolve("ok"),
  getPermissionsAsync: Promise.resolve({
    status: NOTIFICATIONS_PERMISSIONS_GRANTED,
  }),
  requestPermissionsAsync: Promise.resolve({
    status: NOTIFICATIONS_PERMISSIONS_GRANTED,
  }),
  getExpoPushTokenAsync: Promise.resolve({
    data: NOTIFICATIONS_DEFAULT_FAKE_TOKEN,
  }),
  setNotificationChannelAsync: Promise.resolve(),
};

type WithCapturedNotifications = {
  captured: NotificationContent[];
};

/**
 * Create a fake for INotifications that implements provided behavior,
 * and captures sent notifications.
 */
export const makeNotificationsWithBehavior = (
  behaviorOverride: Partial<Behavior> = {},
): INotifications & WithCapturedNotifications => {
  const behavior = { ...defaultBehavior, ...behaviorOverride };
  const captured: NotificationContent[] = [];

  return {
    captured,

    scheduleNotificationAsync: ({ content }) => {
      captured.push(content);

      return behavior.scheduleNotificationAsync;
    },
    getPermissionsAsync: () => behavior.getPermissionsAsync,
    requestPermissionsAsync: () => behavior.requestPermissionsAsync,
    getExpoPushTokenAsync: () => behavior.getExpoPushTokenAsync,
    setNotificationChannelAsync: () => behavior.setNotificationChannelAsync,
  };
};
