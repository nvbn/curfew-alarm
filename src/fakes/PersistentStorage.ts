import IPersistentStorage from "../dependencies/IPersistentStorage";

export const PERSISTENT_STORAGE_BEHAVIOR_OK = "PERSISTENT_STORAGE_BEHAVIOR_OK";

export const PERSISTENT_STORAGE_BEHAVIOR_ERROR =
  "PERSISTENT_STORAGE_BEHAVIOR_ERROR";

type Behavior =
  | typeof PERSISTENT_STORAGE_BEHAVIOR_OK
  | typeof PERSISTENT_STORAGE_BEHAVIOR_ERROR;

type Options = {
  data: Record<string, string | null>;
  getItem: Behavior;
  setItem: Behavior;
};

const defaultOptions: Omit<Options, "data"> = {
  getItem: PERSISTENT_STORAGE_BEHAVIOR_OK,
  setItem: PERSISTENT_STORAGE_BEHAVIOR_OK,
};

/**
 * Fake implementation of IPersistentStorage with overridable behavior.
 */
export const makePersistentStorageWithDataAndBehavior = (
  overrideOptions: Partial<Options> = {},
): IPersistentStorage => {
  const options = { data: {}, ...defaultOptions, ...overrideOptions };

  return {
    getItem: async (key) => {
      if (options.getItem === PERSISTENT_STORAGE_BEHAVIOR_ERROR) {
        throw new Error("expected fake PersistentStorage error");
      }

      return options.data[key] ?? null;
    },
    setItem: async (key, value) => {
      if (options.setItem === PERSISTENT_STORAGE_BEHAVIOR_ERROR) {
        throw new Error("expected fake PersistentStorage error");
      }

      options.data[key] = value;
    },
  };
};
