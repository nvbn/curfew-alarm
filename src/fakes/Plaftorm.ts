import IPlatform, {
  PLATFORM_OS_ANDROID,
  PLATFORM_OS_IOS,
} from "../dependencies/IPlatform";

/**
 * A fake for IOS platform metadata.
 */
export const makePlatformIOS = (): IPlatform => ({
  OS: PLATFORM_OS_IOS,
});

/**
 * A fake for android platform metadata.
 */
export const makePlatformAndroid = (): IPlatform => ({
  OS: PLATFORM_OS_ANDROID,
});
