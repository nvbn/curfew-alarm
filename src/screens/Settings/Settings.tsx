import { useHeaderHeight } from "@react-navigation/stack";
import React from "react";
import { KeyboardAvoidingView } from "react-native";

import SettingsForm, { TypedItemProps } from "../../components/SettingsForm";
import useNotifications from "../../hooks/useNotifications";
import usePlatformOS from "../../hooks/usePlatformOS";
import useSettings from "../../hooks/useSettings";
import { futureMap } from "../../utils/future";
import i18n from "../../utils/i18n";
import { Time } from "../../utils/time";
import styles from "./styles";

/**
 * A screen that allows to modify app settings.
 */
const Settings = (): JSX.Element => {
  const os = usePlatformOS();
  const [settings, updateSettings] = useSettings();
  const [hasNotificationsPermission, requestNotifications] = useNotifications();
  const headerHeight = useHeaderHeight();

  const options: TypedItemProps[] = [
    {
      id: "enable",
      title: i18n.t("inputTitleEnabled"),
      type: "boolean",
      value: futureMap(settings, (v) => v.enabled),
      onChange: (enabled) => updateSettings({ enabled }),
    },
    {
      id: "enable-notifications",
      title: i18n.t("inputTitleEnableNotifications"),
      type: "boolean",
      value: futureMap(
        hasNotificationsPermission,
        (hasNotificationsPermission) =>
          futureMap(
            settings,
            (settings) =>
              hasNotificationsPermission && settings.notificationsEnabled,
          ),
      ),
      onChange: (notificationsEnabled) => {
        if (notificationsEnabled) {
          // re-request doesn't do any harm
          requestNotifications();
        }

        updateSettings({ notificationsEnabled });
      },
    },
    {
      id: "curfew-start",
      title: i18n.t("inputTitleCurfewStart"),
      type: "time",
      value: futureMap(settings, (v) => v.curfewStart),
      onChange: (curfewStart: Time) => updateSettings({ curfewStart }),
    },
    {
      id: "curfew-end",
      title: i18n.t("inputTitleCurfewEnd"),
      type: "time",
      value: futureMap(settings, (v) => v.curfewEnd),
      onChange: (curfewEnd: Time) => updateSettings({ curfewEnd }),
    },
    {
      id: "go-home",
      title: i18n.t("inputTitleGoHome"),
      type: "number",
      value: futureMap(settings, (v) => v.minutesToGoHome),
      onChange: (minutesToGoHome: number) =>
        updateSettings({ minutesToGoHome }),
    },
  ];

  return (
    <KeyboardAvoidingView
      style={styles.container}
      keyboardVerticalOffset={headerHeight}
      behavior={os === "ios" ? "padding" : "height"}
    >
      <SettingsForm options={options} os={os} />
    </KeyboardAvoidingView>
  );
};

export default Settings;
