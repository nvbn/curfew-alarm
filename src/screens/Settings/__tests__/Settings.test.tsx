import { render } from "@testing-library/react-native";
import React from "react";

import Notifications from "../../../contexts/Notifications";
import PersistentStorage from "../../../contexts/PersistentStorage";
import Platform from "../../../contexts/Platform";
import { makeNotificationsWithBehavior } from "../../../fakes/Notifications";
import { makePersistentStorageWithDataAndBehavior } from "../../../fakes/PersistentStorage";
import { makePlatformAndroid, makePlatformIOS } from "../../../fakes/Plaftorm";
import Settings from "../Settings";

describe("<Settings />", () => {
  for (const platform of [makePlatformAndroid(), makePlatformIOS()]) {
    test(`can be rendered on ${platform}`, () => {
      const rendered = render(
        <PersistentStorage.Provider
          value={makePersistentStorageWithDataAndBehavior()}
        >
          <Platform.Provider value={platform}>
            <Notifications.Provider value={makeNotificationsWithBehavior()}>
              <Settings />
            </Notifications.Provider>
          </Platform.Provider>
        </PersistentStorage.Provider>,
      );

      expect(rendered.toJSON()).toMatchSnapshot();
    });
  }
});
