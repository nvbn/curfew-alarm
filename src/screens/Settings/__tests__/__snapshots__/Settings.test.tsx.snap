// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Settings /> can be rendered on [object Object] 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#fff",
      "flex": 1,
    }
  }
>
  <RCTScrollView
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View />
  </RCTScrollView>
</View>
`;

exports[`<Settings /> can be rendered on [object Object] 2`] = `
<View
  style={
    Object {
      "backgroundColor": "#fff",
      "flex": 1,
    }
  }
>
  <RCTScrollView
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View />
  </RCTScrollView>
</View>
`;
